File: src/main/java/org/c3lang/intellij/stubs/C3StructDeclarationStub.java
package org.c3lang.intellij.stubs;

import com.intellij.psi.stubs.IStubElementType;
import com.intellij.psi.stubs.StubBase;
import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import org.c3lang.intellij.psi.C3StructBody;
import org.c3lang.intellij.psi.C3StructDeclaration;
import org.c3lang.intellij.psi.FullyQualifiedName;
import org.c3lang.intellij.psi.StructField;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.ArrayList;

public class C3StructDeclarationStub extends StubBase<C3StructDeclaration>
{
	private final @NotNull FullyQualifiedName typeName;
	private final @NotNull List<StructField> fields;

	public C3StructDeclarationStub(
		@Nullable StubElement<?> parent,
		@Nullable IStubElementType<?, ?> elementType,
		@NotNull FullyQualifiedName typeName,
		@NotNull List<StructField> fields)
	{
		super(parent, elementType);
		this.typeName = typeName;
		this.fields = fields;
	}

	public C3StructDeclarationStub(
		@Nullable StubElement<?> parent,
		@Nullable IStubElementType<?, ?> elementType,
		@NotNull C3StructDeclaration psi)
	{
		this(
			parent,
			elementType,
			FullyQualifiedName.from(psi.getTypeName(), psi.getTypeName().getModuleName()),
			collectFields(psi)
		);
	}

	public C3StructDeclarationStub(
		@Nullable StubElement<?> parent,
		@Nullable IStubElementType<?, ?> elementType,
		@NotNull StubInputStream dataStream) throws IOException
	{
		this(
			parent,
			elementType,
			FullyQualifiedName.parse(dataStream.readUTFFast()),
			deserializeStructFields(dataStream)
		);
	}

	private static @NotNull List<StructField> collectFields(@NotNull C3StructDeclaration psi)
	{
		C3StructBody structBody = psi.getStructBody();
		return structBody != null ? StructField.collectFields(structBody, null) : Collections.emptyList();
	}

	private static @NotNull List<StructField> deserializeStructFields(@NotNull StubInputStream dataStream) throws IOException
	{
		int fieldCount = Integer.parseInt(dataStream.readUTFFast());
		List<StructField> result = new ArrayList<>(fieldCount);

		for (int i = 0; i < fieldCount; i++)
		{
			String typeFullName = dataStream.readUTFFast();
			String name = StubStreamExtensions.readNullableUTFFast(dataStream);
			if (name == null) throw new IllegalStateException("Struct field name missing");
			result.add(new StructField(name, FullyQualifiedName.parse(typeFullName)));
		}

		return result;
	}

	public @NotNull FullyQualifiedName getTypeName()
	{
		return typeName;
	}

	public @NotNull List<StructField> getFields()
	{
		return fields;
	}

	public void serialize(@NotNull StubOutputStream stream) throws IOException
	{
		stream.writeUTFFast(typeName.getFullName());
		stream.writeUTFFast(Integer.toString(fields.size()));
		for (StructField field : fields)
		{
			stream.writeUTFFast(field.getType().getFullName());
			StubStreamExtensions.writeNullableUTFFast(stream, field.getName());
		}
	}
}

File: src/main/java/org/c3lang/intellij/stubs/C3StructDeclarationStub.kt
package org.c3lang.intellij.stubs

import com.intellij.psi.stubs.*
import org.c3lang.intellij.psi.C3StructDeclaration
import org.c3lang.intellij.psi.FullyQualifiedName
import org.c3lang.intellij.psi.StructField
import org.c3lang.intellij.psi.StructField.Companion.collectFields

class C3StructDeclarationStub(
    parent: StubElement<*>?,
    elementType: IStubElementType<*, *>?,
    val typeName: FullyQualifiedName,
    val fields: List<StructField>
) : StubBase<C3StructDeclaration>(parent, elementType) {

    constructor(
        parent: StubElement<*>?,
        elementType: IStubElementType<*, *>?,
        psi: C3StructDeclaration,
    ) : this(
        parent = parent,
        elementType = elementType,
        typeName = FullyQualifiedName.from(psi.typeName),
        fields = psi.structBody?.collectFields(null) ?: emptyList()
    )

    constructor(
        parent: StubElement<*>?,
        elementType: IStubElementType<*, *>?,
        dataStream: StubInputStream,
    ) : this(
        parent = parent,
        elementType = elementType,
        typeName = FullyQualifiedName.parse(dataStream.readUTFFast()),
        fields = dataStream.deserializeStructFields()
    )

    fun serialize(stream: StubOutputStream) {
        stream.writeUTFFast(typeName.fullName)
        stream.writeUTFFast(fields.size.toString())
        fields.forEach { field ->
            stream.writeUTFFast(field.type.fullName)
            stream.writeNullableUTFFast(field.name)
        }
    }

    companion object {
        fun StubInputStream.deserializeStructFields(): List<StructField> {
            val fieldCount = readUTFFast().toInt()
            val result = mutableListOf<StructField>()

            (0 until fieldCount).forEach { i ->
                val typeFullName = readUTFFast()
                val name = readNullableUTFFast()

                StructField(
                    checkNotNull(name), FullyQualifiedName.parse(typeFullName)
                )
            }

            return result
        }
    }
}